src/monitor/nec_monitor_main.cpp

```diff
@@ -0,0 +1,125 @@
+/*
+ * This file is part of the lo2s software.
+ * Linux OTF2 sampling
+ *
+ * Copyright (c) 2017,
+ *    Technische Universitaet Dresden, Germany
+ *
+ * lo2s is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * lo2s is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with lo2s.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <lo2s/monitor/nec_monitor_main.hpp>
+
+namespace lo2s
+{
+namespace monitor
+{
+
+int NecMonitorMain::get_nec_device_id_for_thread(Thread thread)
+{
+    // /sys/class/ve/ve0 is not device 0, because that would make too much sense
+    // So look the device id up here
+
+    int real_device_id = -1;
+    for (int i = 0; i < nodeinfo_.total_node_count; i++)
+    {
+        if (!ve_check_pid(nodeinfo_.nodeid[i], thread.as_pid_t()))
+        {
+            real_device_id = nodeinfo_.nodeid[i];
+            break;
+        }
+    }
+    return real_device_id;
+}
+NecMonitorMain::NecMonitorMain(trace::Trace& trace, int device)
+: ThreadedMonitor(trace, fmt::format("VE {}", device)), trace_(trace), device_(device),
+  stopped_(false)
+{
+    ve_node_info(&nodeinfo_);
```

Comment:
check return for errors. It also sets errno: https://github.com/veos-sxarr-NEC/veosinfo_source/blob/master/veosinfo.c